CONFIGURATION.md
# Configuration Format

The floskell configuration file is a `JSON` file.  The basic structure
is as follows:

```json
{
  "language": "Haskell2010",
  "extensions": [],
  "style": "base",
  "formatting": {
    "penalty": {...},
    "layout": {...},
    "indent": {...},
    "align": {...},
    "group": {...},
    "op": {...},
    "options": {...}
    }
  }
}
```

## Parser Settings

* `language` (`Haskell2010`): The Haskell language standard for parsing source files.
  Either `Haskell98` or `Haskell2010`.

  ```json
  {
    "language": "Haskell2010"
  }
  ```

* `extensions` (`[]`): A list of extensions to be enabled or disabled.  Check
  the output of `floskell --help` for the full list of extensions.  To
  disable an extension, prefix the extension with `No`.

  ```json
  {
    "extensions": ["RecordWildCards", "NoTypeFamilies"]
  }
  ```

* `fixities` (`[]`): A list of additional infix declarations to use
  while parsing.  Check the output of `floskell --help` for a list of
  built-in declarations.

  ```json
  {
    "fixities": ["infixr 9 ."]
  }
  ```

## Style Definition

* `style` (`base`): The formatting style to use.  The style can be fine-tuned
  using the `formatting` configuration element.  Check the output of
  `floskell --help` for the full list of styles.

* `formatting` (`{}`): Formatting settings, applied on top of the selected
  style.

### Line Penalty

* `penalty`: Defines the parameters for the penalty function used to
  judge the overall layout.

  * `indent` (`1`): Penalty for each indentation character.
  * `linebreak` (`10`): Penalty for each line.
  * `max-line-length` (`80`): Character limit for each line.
  * `overfull` (`10`): Penalty for each character over the line limit.
  * `overfull-once` (`200`): Penalty for each overfull line.

### Layout Rules

* `layout`: Defines the accepted layout choices for all configurable
  language constructs.

  Allowed values are `flex`, `vertical`, or `try-oneline`.

  * `app` (`flex`): Function application
  * `con-decls` (`flex`): Data constructor declarations
  * `declaration` (`flex`): Declarations
  * `export-spec-list` (`flex`): Module export list
  * `if` (`flex`): If-then-else
  * `import-spec-list` (`flex`): Module import lists
  * `infix-app` (`flex`): Infix operator application
  * `let` (`flex`): Let-in
  * `list-comp` (`flex`): List comprehensions
  * `record` (`flex`): Record declaration, construction, and update
  * `type` (`flex`): Types in signatures

### Indentation Rules

* `indent`: Defines the indentation choices for all configurable
  language constructs.

  Allowed values (except for `onside`, `deriving`, `where`) are
  `align`, `indent-by n`, and `align-or-indent-by n`.

  Allowed values for `onside` are all integers greater than zero.

  Allowed values for `deriving` and `where` are all integers greater
  than or equal to zero.

  * `onside` (`4`): Onside indentation
  * `deriving` (`4`): Indentation for deriving lists
  * `where` (`2`): Indentation for where keyword after functions
  * `case` (`indent-by 4`): Case analysis
  * `class` (`indent-by 4`): Class and instance declarations
  * `do` (`indent-by 4`): Do and recursive do blocks
  * `export-spec-list` (`indent-by 4`): Module export list
  * `if` (`indent-by 4`): If-then-else
  * `import-spec-list` (`indent-by 4`): Module import lists
  * `let` (`indent-by 4`): Let bindings
  * `let-binds` (`indent-by 4`): Let bindings
  * `let-in` (`indent-by 4`): Expression in let
  * `multi-if` (`indent-by 4`): Guards and multi-way if
  * `typesig` (`indent-by 4`): Type signatures
  * `where-binds` (`indent-by 2`): Bindings in where block after functions

### Tabstop Alignment

* `align`: Defines the enabled tabstop alignments and their limitations.

  Allowed values, except for `limits`, are `true` and `false`.

  The item `limits` must be a list with two integers, the first
  integer is a number of character (`absolute limit`), the second a
  percentage (`relative limit`).  Alignment will be disabled if the
  length of any inserted padding is larger than `absolute limit` and
  is larger than `relative limit` of the maximum field length (tabstop
  column minus indentation).

  * `limits` (`[10, 25]`): Alignment limits
  * `case` (`false`): Case patterns
  * `class` (`false`): Declarations inside class and instance declarations
  * `import-module` (`false`): Names of imported modules
  * `import-spec` (`false`): Module import lists
  * `let-binds` (`false`): Let bindings
  * `matches` (`false`): Function match clauses
  * `record-fields` (`false`): Record field assignments and type signatures
  * `where` (`false`): Declarations in where bindings for functions

### Whitespace Rules

* `group`: Defines whitespace rules for parentheses, brackets,
  braces, and other enclosing punctuation characters.

  The `group` item is a map from enclosing punctuation (identified
  by the opening characters) and optional context to a whitespace
  configuration. The key can be either

  * `default`: Defines the fallback whitespace configuration,

  * `* in <context>`: Defines the default whitespace configuration
    within the given context,

  * `<punct> in *`: Defines the default whitespace configuration for
    the given punctuation, or

  * `<punct> in <context>`: Defines the whitespace configuration for
    the given punctuation within the given context,

  where context is one of `declaration`, `type`, `pattern`,
  `expression`, or `other`.

  The value is an object with the following fields:

  * `spaces`: Where to insert spaces.
  * `linebreaks`: Where to insert linebreaks.
  * `force-linebreak`: Whether to enforce vertical layout.

  An example:

  ```haskell
  {
    "group": {
      "default": {
        "spaces": "none",
        "linebreaks": "after",
        "force-linebreak": false
      },
      "(": {
        "spaces": "both",
        "linebreaks": "after",
        "force-linebreak": false
      },
      "( in pattern": {
        "spaces": "none",
        "linebreaks": "none",
        "force-linebreak": false
      }
    }
  }
  ```

* `op`: Defines whitespacce rules for infix operators.

  See `group` above.

### Other Style Options

* `options`: Defines additional formatting rules for module heads.

  * `align-sum-type-decl` (`false`): Whether to align `=` with `|` in the
    declaration of sum types.
  * `preserve-vertical-space` (`false`): Whether to preserve additional vertical
    space between declarations, statements, and a few other places.
  * `decl-no-blank-lines` (`[]`): Where to (not) insert blank lines between
    declarations (`module`, `class`, `instance`, `where`).
  * `sort-import-lists` (`false`): Whether to sort import statements by the name
    of the imported module.
  * `sort-imports` (`false`): How to sort import lists (see below).
  * `sort-pragmas` (`false`): Whether to sort module pragmas (`LANGUAGE`,
    `OPTION`, and `ANN`).
  * `flexible-oneline` (`false`): Allow `do`, `mdo`, `case ... of`, `\case`, and
    `\... ->` in `try-oneline` layout.
  * `split-language-pragmas` (`false`): Whether to split `LANGUAGE` pragmas.
  * `align-let-binds-and-in-expr` (`false`): Whether to align `let` bindings and
    the `in` expression, in a let expression.

#### Sorting Imports

There are three import sorting modes:

  * No sorting (`sort-imports: false`).  This mode will retain the
    order of imports and any blank lines between imports.

  * Sort imports and group by the first component of the module path
    (`sort-imports: true`): Data.Maybe goes with Data.Either,
    Control.Monad goes with Control.Applicative.

    This mode is equivalent to:

    ``` json
    "sort-imports": [
      {
        "prefixes": [""],
        "order": "grouped"
      }
    ]
    ```

  * Configurable grouping and sorting (`sort-imports:
    [...]`). `sort-imports` is a list of group declarations, where
    each group defines a set of module prefixes and an order.

    Each import is assigned to a group based on the longest matching
    prefix. Imports without a match are assigned to an implicit final
    group with order `keep`. You can manually collect all left-over
    imports by specifying a group with an empty prefix.

    The order can be one of `keep`, `sorted`, or `grouped`.  `keep`
    retains the manual order of imports of the input file, `sorted`
    will sort imports within the group, and `grouped` will sort and
    then group imports within the group based on the first component
    after the common prefix of the group's prefixes.

    For example:

    ``` json
    "sort-imports": [
      {
        "prefixes": [ "" ],
        "order": "grouped"
      },
      {
        "prefixes": [ "Floskell" ],
        "order": "sorted"
      },
      {
        "prefixes": [ "Test" ],
        "order": "grouped"
      },
      {
        "prefixes": [ "Prelude" ],
        "order": "sorted"
      }
    ]
    ```

    This configuration will first place all 'generic' imports, sorted
    and grouped on the top-level module name component. E.g. all of
    "Control", then "Data", etc.

    After that, there is a sorted section of 'local' imports
    (everything within "Floskell"), without any subdivision.

    Following are all imports from "Test", again sorted and grouped
    based on the first component after "Test".

    Lastly, any imports of "Prelude".


CHANGELOG.md
# Floskell 0.10.5 (2020-09-21)

* Fix indentation of nested list and tuple expressions
* Fix formatting of slightly overlong module headers
* Fix formatting of promoted types

# Floskell 0.10.4 (2020-08-05)

* Fix formatting of type applications
* Fix for comments hopping over 'where' keyword
* Add safety whitespace override for @ in patterns

# Floskell 0.10.3 (2020-05-17)

* Updated to haskell-src-exts-1.23.0, with support for Block Arguments
  and Arrow Brackets
* Compatible with GHC-8.10.1

# Floskell 0.10.2 (2019-11-11)

* Updated to haskell-src-exts-1.22.0, with support for Template
  Haskell typed splices and quotations
* Compatible with GHC-8.8.1

# Floskell 0.10.1 (2019-05-25)

* Fix for broken off-side rule with multi-line lists
* Fix for increasing comment indentation before instance declarations

# Floskell 0.10.0 (2019-05-03)

* Updated to haskell-src-exts-1.21.0, with support for `DerivingVia` and `TypeInType`
* Support for custom fixity declarations (with common libraries built in)
* More control over sorting and formatting of imports
* More control over formatting of type signatures
* New option to align `=` with `|` in data declarations
* New option to allow `do`, `case`, and lambda in `try-oneline` layouts
* New option to avoid vertical space between declarations
* New option to tabstop-align right-hand-side of function match clauses
* Improved handling of CPP directives
* Improved positioning of comments
* Better error messages for JSON syntax errors in configuration file
* Many formatting fixes

Incompatible changes:

* `formatting/layout/typesig` has been removed in favor of
  `formatting/layout/type`

# Floskell 0.9.0 (2019-01-25)

* Initial release


src/Floskell/Config.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}

module Floskell.Config
    ( Indent(..)
    , LayoutContext(..)
    , Location(..)
    , WsLoc(..)
    , Whitespace(..)
    , Layout(..)
    , ConfigMapKey(..)
    , ConfigMap(..)
    , PenaltyConfig(..)
    , AlignConfig(..)
    , IndentConfig(..)
    , LayoutConfig(..)
    , OpConfig(..)
    , GroupConfig(..)
    , ImportsGroupOrder(..)
    , ImportsGroup(..)
    , SortImportsRule(..)
    , DeclarationConstruct(..)
    , OptionConfig(..)
    , Config(..)
    , defaultConfig
    , safeConfig
    , cfgMapFind
    , cfgOpWs
    , cfgGroupWs
    , inWs
    , wsSpace
    , wsLinebreak
    ) where

import           Data.Aeson
                 ( FromJSON(..), ToJSON(..), genericParseJSON, genericToJSON )
import qualified Data.Aeson         as JSON
import           Data.Aeson.Types   as JSON
                 ( Options(..), camelTo2, typeMismatch )
import           Data.ByteString    ( ByteString )
import           Data.Default       ( Default(..) )
import qualified Data.HashMap.Lazy  as HashMap
import           Data.Map.Strict    ( Map )
import qualified Data.Map.Strict    as Map
import           Data.Set           ( Set )
import qualified Data.Set           as Set
import qualified Data.Text          as T
import qualified Data.Text.Encoding as T ( decodeUtf8, encodeUtf8 )

import           GHC.Generics

data Indent = Align | IndentBy !Int | AlignOrIndentBy !Int
    deriving ( Eq, Ord, Show, Generic )

data LayoutContext = Declaration | Type | Pattern | Expression | Other
    deriving ( Eq, Ord, Bounded, Enum, Show, Generic )

data Location = Before | After
    deriving ( Eq, Ord, Bounded, Enum, Show, Generic )

data WsLoc = WsNone | WsBefore | WsAfter | WsBoth
    deriving ( Eq, Ord, Bounded, Enum, Show, Generic )

data Whitespace = Whitespace { wsSpaces         :: !WsLoc
                             , wsLinebreaks     :: !WsLoc
                             , wsForceLinebreak :: !Bool
                             }
    deriving ( Show, Generic )

data Layout = Flex | Vertical | TryOneline
    deriving ( Eq, Ord, Bounded, Enum, Show, Generic )

data ConfigMapKey = ConfigMapKey !(Maybe ByteString) !(Maybe LayoutContext)
    deriving ( Eq, Ord, Show )

data ConfigMap a =
    ConfigMap { cfgMapDefault :: !a, cfgMapOverrides :: !(Map ConfigMapKey a) }
    deriving ( Generic )

data PenaltyConfig = PenaltyConfig { penaltyMaxLineLength :: !Int
                                   , penaltyLinebreak     :: !Int
                                   , penaltyIndent        :: !Int
                                   , penaltyOverfull      :: !Int
                                   , penaltyOverfullOnce  :: !Int
                                   }
    deriving ( Generic )

instance Default PenaltyConfig where
    def = PenaltyConfig { penaltyMaxLineLength = 80
                        , penaltyLinebreak     = 100
                        , penaltyIndent        = 1
                        , penaltyOverfull      = 10
                        , penaltyOverfullOnce  = 200
                        }

data AlignConfig =
    AlignConfig { cfgAlignLimits       :: !(Int, Int)
                , cfgAlignCase         :: !Bool
                , cfgAlignClass        :: !Bool
                , cfgAlignImportModule :: !Bool
                , cfgAlignImportSpec   :: !Bool
                , cfgAlignLetBinds     :: !Bool
                , cfgAlignMatches      :: !Bool
                , cfgAlignRecordFields :: !Bool
                , cfgAlignWhere        :: !Bool
                }
    deriving ( Generic )

instance Default AlignConfig where
    def = AlignConfig { cfgAlignLimits       = (10, 25)
                      , cfgAlignCase         = False
                      , cfgAlignClass        = False
                      , cfgAlignImportModule = False
                      , cfgAlignImportSpec   = False
                      , cfgAlignLetBinds     = False
                      , cfgAlignMatches      = False
                      , cfgAlignRecordFields = False
                      , cfgAlignWhere        = False
                      }

data IndentConfig =
    IndentConfig { cfgIndentOnside :: !Int
                 , cfgIndentDeriving :: !Int
                 , cfgIndentWhere :: !Int
                 , cfgIndentApp :: !Indent
                 , cfgIndentCase :: !Indent
                 , cfgIndentClass :: !Indent
                 , cfgIndentDo :: !Indent
                 , cfgIndentExportSpecList :: !Indent
                 , cfgIndentIf :: !Indent
                 , cfgIndentImportSpecList :: !Indent
                 , cfgIndentLet :: !Indent
                 , cfgIndentLetBinds :: !Indent
                 , cfgIndentLetIn :: !Indent
                 , cfgIndentMultiIf :: !Indent
                 , cfgIndentTypesig :: !Indent
                 , cfgIndentWhereBinds :: !Indent
                 }
    deriving ( Generic )

instance Default IndentConfig where
    def = IndentConfig { cfgIndentOnside = 4
                       , cfgIndentDeriving = 4
                       , cfgIndentWhere = 2
                       , cfgIndentApp = IndentBy 4
                       , cfgIndentCase = IndentBy 4
                       , cfgIndentClass = IndentBy 4
                       , cfgIndentDo = IndentBy 4
                       , cfgIndentExportSpecList = IndentBy 4
                       , cfgIndentIf = IndentBy 4
                       , cfgIndentImportSpecList = IndentBy 4
                       , cfgIndentLet = IndentBy 4
                       , cfgIndentLetBinds = IndentBy 4
                       , cfgIndentLetIn = IndentBy 4
                       , cfgIndentMultiIf = IndentBy 4
                       , cfgIndentTypesig = IndentBy 4
                       , cfgIndentWhereBinds = IndentBy 2
                       }

data LayoutConfig =
    LayoutConfig { cfgLayoutApp :: !Layout
                 , cfgLayoutConDecls :: !Layout
                 , cfgLayoutDeclaration :: !Layout
                 , cfgLayoutExportSpecList :: !Layout
                 , cfgLayoutIf :: !Layout
                 , cfgLayoutImportSpecList :: !Layout
                 , cfgLayoutInfixApp :: !Layout
                 , cfgLayoutLet :: !Layout
                 , cfgLayoutListComp :: !Layout
                 , cfgLayoutRecord :: !Layout
                 , cfgLayoutType :: !Layout
                 }
    deriving ( Generic )

instance Default LayoutConfig where
    def = LayoutConfig { cfgLayoutApp = Flex
                       , cfgLayoutConDecls = Flex
                       , cfgLayoutDeclaration = Flex
                       , cfgLayoutExportSpecList = Flex
                       , cfgLayoutIf = Flex
                       , cfgLayoutImportSpecList = Flex
                       , cfgLayoutInfixApp = Flex
                       , cfgLayoutLet = Flex
                       , cfgLayoutListComp = Flex
                       , cfgLayoutRecord = Flex
                       , cfgLayoutType = Flex
                       }

newtype OpConfig = OpConfig { unOpConfig :: ConfigMap Whitespace }
    deriving ( Generic )

instance Default OpConfig where
    def =
        OpConfig ConfigMap { cfgMapDefault   = Whitespace WsBoth WsBefore False
                           , cfgMapOverrides = Map.empty
                           }

newtype GroupConfig = GroupConfig { unGroupConfig :: ConfigMap Whitespace }
    deriving ( Generic )

instance Default GroupConfig where
    def = GroupConfig ConfigMap { cfgMapDefault   =
                                      Whitespace WsBoth WsAfter False
                                , cfgMapOverrides = Map.empty
                                }

data ImportsGroupOrder =
    ImportsGroupKeep | ImportsGroupSorted | ImportsGroupGrouped
    deriving ( Generic )

data ImportsGroup = ImportsGroup { importsPrefixes :: ![String]
                                 , importsOrder    :: !ImportsGroupOrder
                                 }
    deriving ( Generic )

data SortImportsRule =
    NoImportSort | SortImportsByPrefix | SortImportsByGroups ![ImportsGroup]

data DeclarationConstruct = DeclModule | DeclClass | DeclInstance | DeclWhere
    deriving ( Eq, Ord, Generic )

data OptionConfig =
    OptionConfig { cfgOptionSortPragmas            :: !Bool
                 , cfgOptionSplitLanguagePragmas   :: !Bool
                 , cfgOptionSortImports            :: !SortImportsRule
                 , cfgOptionSortImportLists        :: !Bool
                 , cfgOptionAlignSumTypeDecl       :: !Bool
                 , cfgOptionFlexibleOneline        :: !Bool
                 , cfgOptionPreserveVerticalSpace  :: !Bool
                 , cfgOptionDeclNoBlankLines       :: !(Set DeclarationConstruct)
                 , cfgOptionAlignLetBindsAndInExpr :: !Bool
                 }
    deriving ( Generic )

instance Default OptionConfig where
    def = OptionConfig { cfgOptionSortPragmas            = False
                       , cfgOptionSplitLanguagePragmas   = False
                       , cfgOptionSortImports            = NoImportSort
                       , cfgOptionSortImportLists        = False
                       , cfgOptionAlignSumTypeDecl       = False
                       , cfgOptionFlexibleOneline        = False
                       , cfgOptionPreserveVerticalSpace  = False
                       , cfgOptionDeclNoBlankLines       = Set.empty
                       , cfgOptionAlignLetBindsAndInExpr = False
                       }

data Config = Config { cfgPenalty :: !PenaltyConfig
                     , cfgAlign   :: !AlignConfig
                     , cfgIndent  :: !IndentConfig
                     , cfgLayout  :: !LayoutConfig
                     , cfgOp      :: !OpConfig
                     , cfgGroup   :: !GroupConfig
                     , cfgOptions :: !OptionConfig
                     }
    deriving ( Generic )

instance Default Config where
    def = Config { cfgPenalty = def
                 , cfgAlign   = def
                 , cfgIndent  = def
                 , cfgLayout  = def
                 , cfgOp      = def
                 , cfgGroup   = def
                 , cfgOptions = def
                 }

defaultConfig :: Config
defaultConfig =
    def { cfgOp    = OpConfig ((unOpConfig def) { cfgMapOverrides =
                                                      Map.fromList opWsOverrides
                                                })
        , cfgGroup = GroupConfig ((unGroupConfig def) { cfgMapOverrides = Map.fromList groupWsOverrides
                                                      })
        }
  where
    opWsOverrides =
        [ (ConfigMapKey (Just ",") Nothing, Whitespace WsAfter WsBefore False)
        , ( ConfigMapKey (Just "record") Nothing
          , Whitespace WsAfter WsAfter False
          )
        , ( ConfigMapKey (Just ".") (Just Type)
          , Whitespace WsAfter WsAfter False
          )
        ]

    groupWsOverrides =
        [ (ConfigMapKey (Just "[") (Just Type), Whitespace WsBoth WsNone False)
        ]

safeConfig :: Config -> Config
safeConfig cfg = cfg { cfgGroup = group, cfgOp = op }
  where
    group = GroupConfig $
        updateOverrides (unGroupConfig $ cfgGroup cfg)
                        [ ("(#", Expression, WsBoth), ("(#", Pattern, WsBoth) ]

    op = OpConfig $
        updateOverrides (unOpConfig $ cfgOp cfg)
                        [ (".", Expression, WsBoth), ("@", Pattern, WsNone) ]

    updateOverrides config overrides =
        config { cfgMapOverrides =
                     foldl (updateWs config) (cfgMapOverrides config) overrides
               }

    updateWs config m (key, ctx, ws) =
        Map.insertWith (flip const)
                       (ConfigMapKey (Just key) (Just ctx))
                       (cfgMapFind ctx key config) { wsSpaces = ws }
                       m

cfgMapFind :: LayoutContext -> ByteString -> ConfigMap a -> a
cfgMapFind ctx key ConfigMap{..} =
    let value = cfgMapDefault
        value' = Map.findWithDefault value
                                     (ConfigMapKey Nothing (Just ctx))
                                     cfgMapOverrides
        value'' = Map.findWithDefault value'
                                      (ConfigMapKey (Just key) Nothing)
                                      cfgMapOverrides
        value''' = Map.findWithDefault value''
                                       (ConfigMapKey (Just key) (Just ctx))
                                       cfgMapOverrides
    in
        value'''

cfgOpWs :: LayoutContext -> ByteString -> OpConfig -> Whitespace
cfgOpWs ctx op = cfgMapFind ctx op . unOpConfig

cfgGroupWs :: LayoutContext -> ByteString -> GroupConfig -> Whitespace
cfgGroupWs ctx op = cfgMapFind ctx op . unGroupConfig

inWs :: Location -> WsLoc -> Bool
inWs _ WsBoth = True
inWs Before WsBefore = True
inWs After WsAfter = True
inWs _ _ = False

wsSpace :: Location -> Whitespace -> Bool
wsSpace loc ws = loc `inWs` wsSpaces ws

wsLinebreak :: Location -> Whitespace -> Bool
wsLinebreak loc ws = loc `inWs` wsLinebreaks ws

------------------------------------------------------------------------
readMaybe :: Read a => String -> Maybe a
readMaybe str = case reads str of
    [ (x, "") ] -> Just x
    _ -> Nothing

enumOptions :: Int -> Options
enumOptions n =
    JSON.defaultOptions { constructorTagModifier = JSON.camelTo2 '-' . drop n }

recordOptions :: Int -> Options
recordOptions n =
    JSON.defaultOptions { fieldLabelModifier = JSON.camelTo2 '-' . drop n
                        , unwrapUnaryRecords = True
                        }

instance ToJSON Indent where
    toJSON i = JSON.String $ case i of
        Align -> "align"
        IndentBy x -> "indent-by " `T.append` T.pack (show x)
        AlignOrIndentBy x -> "align-or-indent-by " `T.append` T.pack (show x)

instance FromJSON Indent where
    parseJSON v@(JSON.String t) = maybe (JSON.typeMismatch "Indent" v) return $
        if t == "align"
        then Just Align
        else if "indent-by " `T.isPrefixOf` t
             then IndentBy <$> readMaybe (T.unpack $ T.drop 10 t)
             else if "align-or-indent-by " `T.isPrefixOf` t
                  then AlignOrIndentBy <$> readMaybe (T.unpack $ T.drop 19 t)
                  else Nothing

    parseJSON v = JSON.typeMismatch "Indent" v

instance ToJSON LayoutContext where
    toJSON = genericToJSON (enumOptions 0)

instance FromJSON LayoutContext where
    parseJSON = genericParseJSON (enumOptions 0)

instance ToJSON WsLoc where
    toJSON = genericToJSON (enumOptions 2)

instance FromJSON WsLoc where
    parseJSON = genericParseJSON (enumOptions 2)

instance ToJSON Whitespace where
    toJSON = genericToJSON (recordOptions 2)

instance FromJSON Whitespace where
    parseJSON = genericParseJSON (recordOptions 2)

instance ToJSON Layout where
    toJSON = genericToJSON (enumOptions 0)

instance FromJSON Layout where
    parseJSON = genericParseJSON (enumOptions 0)

layoutToText :: LayoutContext -> T.Text
layoutToText Declaration = "declaration"
layoutToText Type = "type"
layoutToText Pattern = "pattern"
layoutToText Expression = "expression"
layoutToText Other = "other"

textToLayout :: T.Text -> Maybe LayoutContext
textToLayout "declaration" = Just Declaration
textToLayout "type" = Just Type
textToLayout "pattern" = Just Pattern
textToLayout "expression" = Just Expression
textToLayout "other" = Just Other
textToLayout _ = Nothing

keyToText :: ConfigMapKey -> T.Text
keyToText (ConfigMapKey Nothing Nothing) = "default"
keyToText (ConfigMapKey (Just n) Nothing) = T.decodeUtf8 n
keyToText (ConfigMapKey Nothing (Just l)) = "* in " `T.append` layoutToText l
keyToText (ConfigMapKey (Just n) (Just l)) =
    T.decodeUtf8 n `T.append` " in " `T.append` layoutToText l

textToKey :: T.Text -> Maybe ConfigMapKey
textToKey t = case T.splitOn " in " t of
    [ "default" ] -> Just (ConfigMapKey Nothing Nothing)
    [ "*", "*" ] -> Just (ConfigMapKey Nothing Nothing)
    [ name ] -> Just (ConfigMapKey (Just (T.encodeUtf8 name)) Nothing)
    [ name, "*" ] -> Just (ConfigMapKey (Just (T.encodeUtf8 name)) Nothing)
    [ "*", layout ] -> ConfigMapKey Nothing . Just <$> textToLayout layout
    [ name, layout ] -> ConfigMapKey (Just (T.encodeUtf8 name)) . Just
        <$> textToLayout layout
    _ -> Nothing

instance ToJSON a => ToJSON (ConfigMap a) where
    toJSON ConfigMap{..} = toJSON $ Map.insert "default" cfgMapDefault $
        Map.mapKeys keyToText cfgMapOverrides

instance FromJSON a => FromJSON (ConfigMap a) where
    parseJSON value = do
        o <- parseJSON value
        cfgMapDefault <- maybe (fail "Missing key: default") return $
            HashMap.lookup "default" o
        cfgMapOverrides <- either fail (return . Map.fromList) $ mapM toKey $
            HashMap.toList $ HashMap.delete "default" o
        return ConfigMap { .. }
      where
        toKey (k, v) = case textToKey k of
            Just k' -> Right (k', v)
            Nothing -> Left ("Invalid key: " ++ T.unpack k)

instance ToJSON PenaltyConfig where
    toJSON = genericToJSON (recordOptions 7)

instance FromJSON PenaltyConfig where
    parseJSON = genericParseJSON (recordOptions 7)

instance ToJSON AlignConfig where
    toJSON = genericToJSON (recordOptions 8)

instance FromJSON AlignConfig where
    parseJSON = genericParseJSON (recordOptions 8)

instance ToJSON IndentConfig where
    toJSON = genericToJSON (recordOptions 9)

instance FromJSON IndentConfig where
    parseJSON = genericParseJSON (recordOptions 9)

instance ToJSON LayoutConfig where
    toJSON = genericToJSON (recordOptions 9)

instance FromJSON LayoutConfig where
    parseJSON = genericParseJSON (recordOptions 9)

instance ToJSON OpConfig where
    toJSON = genericToJSON (recordOptions 0)

instance FromJSON OpConfig where
    parseJSON = genericParseJSON (recordOptions 0)

instance ToJSON GroupConfig where
    toJSON = genericToJSON (recordOptions 0)

instance FromJSON GroupConfig where
    parseJSON = genericParseJSON (recordOptions 0)

instance ToJSON ImportsGroupOrder where
    toJSON = genericToJSON (enumOptions 12)

instance FromJSON ImportsGroupOrder where
    parseJSON = genericParseJSON (enumOptions 12)

instance ToJSON ImportsGroup where
    toJSON = genericToJSON (recordOptions 7)

instance FromJSON ImportsGroup where
    parseJSON x@JSON.Array{} =
        ImportsGroup <$> parseJSON x <*> pure ImportsGroupKeep
    parseJSON x = genericParseJSON (recordOptions 7) x

instance ToJSON SortImportsRule where
    toJSON NoImportSort = toJSON False
    toJSON SortImportsByPrefix = toJSON True
    toJSON (SortImportsByGroups xs) = toJSON xs

instance FromJSON SortImportsRule where
    parseJSON (JSON.Bool False) = return NoImportSort
    parseJSON (JSON.Bool True) = return SortImportsByPrefix
    parseJSON v = SortImportsByGroups <$> parseJSON v

instance ToJSON DeclarationConstruct where
    toJSON = genericToJSON (enumOptions 4)

instance FromJSON DeclarationConstruct where
    parseJSON = genericParseJSON (enumOptions 4)

instance ToJSON OptionConfig where
    toJSON = genericToJSON (recordOptions 9)

instance FromJSON OptionConfig where
    parseJSON = genericParseJSON (recordOptions 9)

instance ToJSON Config where
    toJSON = genericToJSON (recordOptions 3)

instance FromJSON Config where
    parseJSON = genericParseJSON (recordOptions 3)
